File: risc0/zkp/verify/fri.cpp
// Copyright 2022 Risc0, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "risc0/zkp/verify/fri.h"

#include <memory>

#include "risc0/core/log.h"
#include "risc0/core/util.h"
#include "risc0/zkp/core/constants.h"
#include "risc0/zkp/core/ntt.h"
#include "risc0/zkp/core/rou.h"
#include "risc0/zkp/verify/merkle.h"

namespace risc0 {

namespace {

// Given polynomial coefficents for f(x) of degree D, there are N unique polynomials g_i, of degree
// D/N such that f(x) = sum_i g_i(x^N) x^i.

// Given N coeffients m_i, one can compute a 'folded' polynomial h of degree D/N, such that h(x) =
// sum_i m_i * g_i(x).

// Given an the evaluation of a polynomial f(x) as above over a set of N points x_i, where x_i =
// w^(i*S + j), 0 <= i < N, and w^S*N == 1, and given m_i as per above, we need a way to
// compute h(w^(j*N)).

// Begin with the identity defined by the folding
// f(x) = sum_i g_i(x^N) x^i
//
// Plug is the values of w^(j*S + k), 0 <= j < N (basically reindexing our assumption).
// f(w^(j*S + k)) = sum_i g_i(w^(j*S*N + k*N)) * w^(j*S+k)^i
// f(w^(j*S + k)) = sum_i g_i(w^(k*N)) * w^(j*S+k)^i
//
// Treating w, S, and k as a constants above, and calling
// F_j = f(w^(j*S + k))
// G_i = g_i(w^(k*N))
// We basically get the set of linear relations:
// For all j: F_j = sum_i G_i * w^(i*j*S+k*i)
//
// Rearranging slightly, we have
// For all j: F_j = sum_i w^(i*j*S) w^(k*i) * G_i
//
// This in turn looks like multiplying each element of G_i by w^(k*i) and doing a fourier transform.
// Thus to reverse it (from F_j -> G_i) we do a reverse fourier transform and than an eltwise
// multiply by w^(-k*i).
//
// Finally, we have all the g_i(w^(k*N)), we just need to mix them via m_i and we are done
// h(w^(k*N)) = sum_i m_i G_i
//
// We replace k by j in the above since j is no longer used.  Also, N is fixed to kFriFold, and we
// set m_i = m^i for m='mix' below.

Fp4 foldEval(Fp4* values, Fp4 mix, size_t S, size_t j) {
  size_t N = kFriFold;
  interpolateNTT(values, kFriFold);
  bitReverse(values, kFriFold);
  size_t rootPo2 = log2Ceil(N * S);
  Fp invWK = pow(kRouRev[rootPo2], j);
  Fp mul(1);
  Fp4 tot;
  Fp4 mixPow(1);
  for (size_t i = 0; i < N; i++) {
    tot += values[i] * mul * mixPow;
    mul *= invWK;
    mixPow *= mix;
  }
  return tot;
}

struct VerifyRoundInfo {
  size_t domain;
  MerkleTreeVerifier merkle;
  Fp4 mix;

  VerifyRoundInfo(ReadIOP& iop, size_t inDomain)
      : domain(inDomain / kFriFold)
      , merkle(iop, domain, kFriFold * 4, kQueries)
      , mix(Fp4::random(iop)) {}

  void verifyQuery(ReadIOP& iop, size_t* pos, Fp4* goal) const {
    // Compute which group we are in
    size_t quot = *pos / domain;
    size_t group = *pos % domain;
    // Get the column data
    auto data = merkle.verify(iop, group);
    std::vector<Fp4> data4(kFriFold);
    for (size_t i = 0; i < kFriFold; i++) {
      data4[i] = Fp4(data[0 * kFriFold + i],
                     data[1 * kFriFold + i],
                     data[2 * kFriFold + i],
                     data[3 * kFriFold + i]);
    }
    // Check the existing goal
    REQUIRE(data4[quot] == *goal);
    // Compute the new goal + pos
    *goal = foldEval(data4.data(), mix, domain, group);
    *pos = group;
  }
};

} // namespace

// Verify a FRI proof,
void friVerify(ReadIOP& iop, size_t deg, InnerVerify inner) {
  size_t domain = deg * kInvRate;
  size_t origDomain = domain;
  std::vector<VerifyRoundInfo> rounds;
  // Prep the folding verfiers
  while (deg > kFriMinDegree) {
    rounds.emplace_back(iop, domain);
    domain /= kFriFold;
    deg /= kFriFold;
  }
  // Grab the final coeffs + commit
  std::vector<Fp> finalCoeffs(deg * 4);
  iop.read(finalCoeffs.data(), finalCoeffs.size());
  auto digest = shaHash(finalCoeffs.data(), finalCoeffs.size(), 1, false);
  iop.commit(digest);
  // Get the generator for the final polynomial evaluations
  Fp gen = kRouFwd[log2Ceil(domain)];
  // Do queries
  for (size_t q = 0; q < kQueries; q++) {
    // Get a 'random' index.
    uint32_t rng = iop.generate();
    size_t pos = rng % origDomain;
    // Do the 'inner' verification for this index
    Fp4 goal = inner(iop, pos);
    // Verify the per-round proofs
    for (auto& round : rounds) {
      round.verifyQuery(iop, &pos, &goal);
    }
    // Do final verification
    Fp x = pow(gen, pos);
    Fp4 fx;
    Fp cur(1);
    for (size_t i = 0; i < deg; i++) {
      for (size_t j = 0; j < 4; j++) {
        fx.elems[j] += cur * finalCoeffs[j * deg + i];
      }
      cur *= x;
    }
    REQUIRE(fx == goal);
  }
}

} // namespace risc0

File: risc0/zkp/src/verify/fri.rs
// Copyright 2026 RISC Zero, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use alloc::vec::Vec;
use core::ops::DerefMut;

use risc0_core::field::{Elem, ExtElem, Field, RootsOfUnity};

use super::Verifier;
use crate::{
    core::{
        hash::HashFn,
        log2_ceil,
        ntt::{bit_reverse, interpolate_ntt},
    },
    verify::{merkle::MerkleTreeVerifier, read_iop::ReadIOP, VerificationError},
    FRI_FOLD, FRI_FOLD_PO2, FRI_MIN_DEGREE, INV_RATE, QUERIES,
};

/// VerifyRoundInfo contains the data against which the queries for a particular
/// round are checked. This includes the Merkle tree top row data, as well as
/// the size of the domain of the polynomial, and the mixing parameter.
struct VerifyRoundInfo<'a, F: Field> {
    domain: usize,
    merkle: MerkleTreeVerifier<'a>,
    mix: F::ExtElem,
}

impl<'a, F: Field> VerifyRoundInfo<'a, F> {
    pub fn new(
        iop: &mut ReadIOP<'a, F>,
        hashfn: &dyn HashFn<F>,
        in_domain: usize,
    ) -> Result<Self, VerificationError> {
        let domain = in_domain / FRI_FOLD;
        Ok(VerifyRoundInfo {
            domain,
            merkle: MerkleTreeVerifier::new(
                iop,
                hashfn,
                domain,
                FRI_FOLD * F::ExtElem::EXT_SIZE,
                QUERIES,
            )?,
            mix: iop.random_ext_elem(),
        })
    }
}

impl<'a, F> Verifier<'a, F>
where
    F: Field,
{
    fn verify_query(
        &self,
        round: &mut VerifyRoundInfo<'a, F>,
        pos: &mut usize,
        goal: &mut F::ExtElem,
    ) -> Result<(), VerificationError> {
        let quot = *pos / round.domain;
        let group = *pos % round.domain;
        // Get the column data
        let hashfn = self.suite.hashfn.as_ref();
        let data = round.merkle.verify(self.iop().deref_mut(), hashfn, group)?;
        let mut data_ext: Vec<F::ExtElem> = (0..FRI_FOLD)
            .map(|i| {
                let mut inps = Vec::with_capacity(F::ExtElem::EXT_SIZE);
                for j in 0..F::ExtElem::EXT_SIZE {
                    inps.push(data[j * FRI_FOLD + i]);
                }
                F::ExtElem::from_subelems(inps)
            })
            .collect();
        // Check the existing goal
        if data_ext[quot] != *goal {
            return Err(VerificationError::InvalidProof);
        }
        // Compute the new goal + pos
        let root_po2 = log2_ceil(FRI_FOLD * round.domain);
        let inv_wk = F::Elem::ROU_REV[root_po2].pow(group);

        interpolate_ntt::<F::Elem, F::ExtElem>(&mut data_ext);
        bit_reverse(&mut data_ext);
        *goal = self.poly_eval(&data_ext, round.mix * inv_wk);

        *pos = group;
        Ok(())
    }

    pub fn fri_verify<InnerFn>(&self, mut inner: InnerFn) -> Result<(), VerificationError>
    where
        InnerFn: FnMut(usize) -> Result<F::ExtElem, VerificationError>,
    {
        let mut degree: usize = self.tot_cycles;
        let hashfn = self.suite.hashfn.as_ref();
        let orig_domain = INV_RATE * degree;
        let mut domain = orig_domain;
        // Prep the folding verifiers
        let rounds_capacity = log2_ceil(degree.div_ceil(FRI_FOLD).div_ceil(FRI_FOLD_PO2));
        let mut rounds = Vec::with_capacity(rounds_capacity);
        while degree > FRI_MIN_DEGREE {
            rounds.push(VerifyRoundInfo::new(
                self.iop().deref_mut(),
                hashfn,
                domain,
            )?);
            domain /= FRI_FOLD;
            degree /= FRI_FOLD;
        }
        // We want to minimize reallocation in verify, so make sure we
        // didn't have to reallocate.
        assert!(
            rounds.len() <= rounds_capacity,
            "Did not allocate enough rounds; needed {} for degree {} but only allocated {}",
            rounds.len(),
            degree,
            rounds_capacity
        );
        // Grab the final coeffs + commit
        let final_coeffs = self
            .iop()
            .read_field_elem_slice(F::ExtElem::EXT_SIZE * degree)?;
        let final_digest = hashfn.hash_elem_slice(final_coeffs);
        self.iop().commit(&final_digest);
        // Get the generator for the final polynomial evaluations
        let gen = <F::Elem as RootsOfUnity>::ROU_FWD[log2_ceil(domain)];
        // Do queries
        let mut poly_buf: Vec<F::ExtElem> = Vec::with_capacity(degree);
        for _ in 0..QUERIES {
            let mut pos = self.iop().random_bits(log2_ceil(orig_domain)) as usize;
            // Do the 'inner' verification for this index
            let mut goal = inner(pos)?;
            // Verify the per-round proofs
            for round in &mut rounds {
                self.verify_query(round, &mut pos, &mut goal)?;
            }
            // Do final verification
            let x = gen.pow(pos);

            poly_buf.clear();
            poly_buf.extend((0..degree).map(|i| {
                F::ExtElem::from_subelems(
                    (0..F::ExtElem::EXT_SIZE).map(|j| final_coeffs[j * degree + i]),
                )
            }));
            let fx = self.poly_eval(poly_buf.as_slice(), F::ExtElem::from_subfield(&x));
            if fx != goal {
                return Err(VerificationError::InvalidProof);
            }
        }
        Ok(())
    }
}
